syntax = "proto2";

package PyPacket;

// --------------
// Helper Classes
// --------------
message dummy_msg{
	required uint32 packetNum = 1;
	required string ID = 2;
	required double time = 3;
	required string s = 4;
}

//3 Element Vector
message Vector3 {
	required float x = 1;
	required float y = 2;
	required float z = 3;
}

message Node {
	required string ID = 1;
	required string dataType = 2; 
	required string commType = 3;
	required string address = 4;
	required string port = 5;
	optional string NMID = 6; //ID of the network Manger associated with controlling said node
}

message Subscriber{
	required string id = 1;
	required string datatype = 2;
	required uint32 port = 3;
	required string address = 4;
}

// -------------------------
// Heartbeat Messsage
// -------------------------
message NMHeartBeat{
	required uint32 packetNum = 1;
	required string ID = 2;
	required double time = 3;
	repeated Node nodes = 4;
}

message NodeHeartBeat{
	required uint32 packetNum = 1;
	required string ID = 2;
	required double time = 3;
	repeated Subscriber sub = 4;
}

// --------------
//  State Messages
// --------------

message SixDofState{

	required uint32 packetNum = 1;
	required string ID = 2;
	required double time = 3;

	required Vector3 NED = 4;
	required Vector3 VEL_Iner = 5;
	required Vector3 ATTITUDE = 6;
	required Vector3 OMEGA = 7;
	
	optional Vector3 LLA = 8;
}

message Simple1AKinematicsState{
	required uint32 packetNum = 1;
	required string ID = 2;
	required double time = 3;
}

message SimpleDubinsState{
	required uint32 packetNum = 1;
	required string ID = 2;
	required double time = 3;
}


// ---------------
// Sensor Messages
// ---------------

message SensorLocation{
	required Vector3 LLA = 1;
	required Vector3 ENU = 2;
}

message RFSensorMsg{
	required float path_loss = 1;
	required float rss = 2;
	optional uint32 packetNum = 3;
}

message RFNode{
	required string chanID = 1; //channel ID number
	required float rssi = 2; //RSSI of channel
	required float pl_msr = 3; //calculated "measured" path loss
	required float pl_error = 4; //PL error between predicted and measured
	optional int xgridNum = 5; //x grid number in the prediction field
	optional int ygridNum = 6; //y grid number in the prediction field
}

message RF_Data_Msg{
	required uint32 packetNum = 1;
	required string ID = 2;
	required double time = 3;
	
	required Vector3 lla = 4;
	optional Vector3 attitude = 5;
	optional double airspeed = 6;
	
	repeated RFNode rfNode = 7;
}